### PR TITLE
Uncertainty Calibration

### PR DESCRIPTION
**Pull request by M-R-Schaefer**
- Added post-hoc calibration functionality using uncertainty toolbox as a function transformation for ASE. 
- Further, added an IPS node which makes this available for workflows.
- Added a chunked version of the shallow ensemble force uncertainties to enable the usage of large ensembles.
- Use function transformations in ase calc batch eval

### CODE REVIEW

**Comment by M-R-Schaefer** [comment]
pre-commit.ci autofix

**Comment by M-R-Schaefer** [comment]
pre-commit.ci autofix

**Comment by M-R-Schaefer** [comment]
pre-commit.ci autofix

File: apax/md/ase_calc.py [hunk 1]
```diff
@@ -95,6 +95,7 @@ def unpack_results(results, inputs):
     unpacked_results = []
     for i in range(n_structures):
         single_results = jax.tree_map(lambda x: x[i], results)
+        single_results["energy"] = single_results["energy"].item()
         for k, v in single_results.items():
```
Comment (PythonFZ) [inline apax/md/ase_calc.py:99]: wasn't this already in main, maybe merge main into this branch first?

File: apax/md/ase_calc.py [hunk 2]
```diff
@@ -182,7 +183,7 @@ def initialize(self, atoms):
             model = make_ensemble(model)
 
         for transformation in self.transformations:
-            model = transformation.apply(model, self.n_models)
+            model = transformation.apply(model)
```
Comment (PythonFZ) [inline apax/md/ase_calc.py:186]: I guess this parameter was moved inside the model?

File: apax/md/function_transformations.py [hunk 3]
```diff
@@ -6,22 +6,22 @@
...
-        (E_bias, results), F_bias = gamd_fn(positions, Z, idx, box, offsets)
+        (E_bias, results), F_bias = bias_and_grad_fn(positions, Z, idx, box, offsets)
 
         if "energy_unbiased" not in results.keys():
             results["energy_unbiased"] = results["energy"]
             results["forces_unbiased"] = results["forces"]
 
         results["energy"] = results["energy"] + E_bias
-        results["forces"] = results["forces"] + F_bias
+        results["forces"] = results["forces"] - F_bias
```
Comment (PythonFZ) [inline apax/md/function_transformations.py:24]: :eyes:

File: apax/md/ase_calc.py (hunk 2 above)
Comment (PythonFZ) [reply apax/md/ase_calc.py:186]: Same with the code below.

File: apax/model/gmnn.py [hunk 4]
```diff
@@ -223,13 +242,32 @@ def __call__(
         }
 
         if self.force_variance:
-            forces_ens = -jax.jacrev(self.energy_model)(R, Z, neighbor, box, offsets)
+            if not self.chunk_size:
+                forces_ens = -jax.jacrev(self.energy_model)(R, Z, neighbor, box, offsets)
+            else:
+                with jax.ensure_compile_time_eval():
+                    assert n_ens % self.chunk_size == 0
+
```
Comment (PythonFZ) [inline apax/model/gmnn.py:250]: assertion should not be used, can this be a value or runtime error?

File: apax/nodes/model.py [hunk 5]
```diff
@@ -204,9 +219,96 @@ def get_calculator(self, **kwargs) -> ase.calculators.calculator.Calculator:
...
+    optimizer_bounds: t.Tuple[float, float] = zntrack.params((1e-2, 1e2))
+
+    transformations: t.Optional[list[dict[str, dict]]] = zntrack.params(None)
+
+    nl_skin: float = zntrack.params(0.5)
+
+    metrics = zntrack.metrics()
+
+    e_factor: float = zntrack.outs()
+    f_factor: float = zntrack.outs()
```
Comment (PythonFZ) [inline apax/nodes/model.py:268]: Would it be sufficient to just store them in `metrics`?

File: apax/md/function_transformations.py [hunk 6]
```diff
@@ -96,7 +97,48 @@ def gamd_energy(positions, Z, idx, box, offsets):
...
+            )
+
+            Emean = results["energy"]
+            Ei = results["energy_ensemble"]
+            results["energy_ensemble"] = Emean + self.energy_factor * (Ei - Emean)
+
+            if "forces_uncertainty" in results.keys():
+                results["forces_uncertainty"] = (
+                    results["forces_uncertainty"] * self.forces_factor
+                )
```
Comment (PythonFZ) [inline apax/md/function_transformations.py:133]: Maybe these keys could be imported from some class that lists all keys used by Apax and also fix https://github.com/apax-hub/apax/issues/308

Doesn't have to be in this PR but should be done soon.

**Review by PythonFZ — APPROVED** [review]
Some small comments to be addressed. As I have no experience in the calibration, I trust your work there. No real testing scenario makes it difficult to review that part.

File: apax/md/ase_calc.py (hunk 1 above)
Comment (M-R-Schaefer) [reply apax/md/ase_calc.py:99]: yeah well, you independently fixed it while the PR was open :sweat_smile:

File: apax/md/function_transformations.py (hunk 6 above)
Comment (M-R-Schaefer) [reply apax/md/function_transformations.py:133]: For what purpose would you import them? Yeah the fix is coming.

File: apax/model/gmnn.py (hunk 4 above)
Comment (M-R-Schaefer) [reply apax/model/gmnn.py:250]: yes

File: apax/md/ase_calc.py (hunk 2 above)
Comment (M-R-Schaefer) [reply apax/md/ase_calc.py:186]: it's now detected in the transformation itself.

File: apax/md/function_transformations.py (hunk 3 above)
Comment (M-R-Schaefer) [reply apax/md/function_transformations.py:24]: Forces are the negative gradient :upside_down_face:

File: apax/nodes/model.py (hunk 5 above)
Comment (M-R-Schaefer) [reply apax/nodes/model.py:268]: I guess so

File: apax/md/function_transformations.py (hunk 6 above)
Comment (M-R-Schaefer) [reply apax/md/function_transformations.py:133]: Ah I guess, we could have an ENERGY_KEY imported from somewhere, yeah but I feel like this is not necessary
